xarray/core/combine.py

```diff
@@ -369,34 +370,178 @@ def _auto_concat(datasets, dim=None, data_vars='all', coords='different'):
 _CONCAT_DIM_DEFAULT = '__infer_concat_dim__'
 
 
+def _infer_concat_order_from_nested_list(datasets, concat_dims):
+
+    combined_ids = _infer_tile_ids_from_nested_list(datasets, [], {})
+
+    # Currently if concat_dims is not supplied then _auto_concat attempts to deduce it on every call
+    # TODO might be faster in this case to just work out the concat_dims once here
+    tile_id, ds = list(combined_ids.items())[0]
+    n_dims = len(tile_id)
+    if concat_dims is None or concat_dims == _CONCAT_DIM_DEFAULT:
+        concat_dims = [_CONCAT_DIM_DEFAULT]*n_dims
+    else:
+        if len(concat_dims) != n_dims:
+            raise ValueError("concat_dims has length " + str(len(concat_dims))
+                             + " but the datasets passed are nested in a " +
+                             str(n_dims) + "-dimensional structure")
+
+    return combined_ids, concat_dims
+
+
+def _infer_tile_ids_from_nested_list(entry, current_pos, combined_tile_ids):
+    """
+    Given a list of lists (of lists...) of datasets, returns a dictionary
+    with the index of each dataset in the nested list structure as the key.
+
+    Recursively traverses the given structure, while keeping track of the
+    current position.
+
+    Parameters
+    ----------
+    entry : list[list[xarray.Dataset, xarray.Dataset, ...]]
+        List of lists of arbitrary depth, containing datasets in the order they
+        are to be concatenated.
+
+    Returns
+    -------
+    combined_tile_ids : dict[tuple(int, ...), xarray.Dataset]
+    """
+
+    from .dataset import Dataset
+
+    if isinstance(entry, list):
+        # Dive down tree and recursively open the next list
```

Comment:
I like this recursive approach, but I would suggest doing all the logic with immutable tuples. Pure functions are easier to reason about and less error prone.

Also, it's generally considered bad practice for a function to both mutate it's arguments and have a return value -- this makes it a less obvious whether a function uses mutation or not. You'll notice that all builtin functions and methods in Python either mutate their inputs (e.g., `list.append()`) or have a return value, but never do both.

I think this function could be really cleanly written as an iterator, e.g.,
```python
def _infer_tile_ids_from_nested_list(entry, current_pos):
    if isinstance(entry, list):
        for i, item in enumerate(entry):
            for result in _infer_tile_ids_from_nested_list(item, current_pos + (i,)):
                yield result
    else:
        yield current_pos, entry
``` 
Then you call it wrapping it with the `dict` constructor, e.g.,`combined_ids = dict(_infer_tile_ids_from_nested_list(datasets, ()))`

Reply:
Okay thanks, I didn't know about the mutate/return convention, but that makes sense.

Reply:
I came up with a better way to do this which involves another argument `ids` being passed the internal implementation of `auto_combine`, depending on whether or not the combined_ids have already been sorted. A prototype version is in the latest commits.